restapi/src/main/java/io/stargate/web/docsapi/service/DocumentService.java

```diff
@@ -390,6 +391,122 @@ private Object convertToBackendBooleanValue(boolean value, boolean numericBoolea
     return ImmutablePair.of(bindVariableList, firstLevelKeys);
   }
 
+  private String convertToJsonPtr(String path) {
+    return "/" + path.replaceAll(PERIOD_PATTERN.pattern(), "/").replaceAll("\\[(\\d+)\\]", "$1");
+  }
+
+  private DocumentDB maybeCreateTableAndIndexes(
+      Db dbFactory,
+      DocumentDB db,
+      String keyspace,
+      String collection,
+      Map<String, String> headers,
+      String authToken)
+      throws UnauthorizedException {
+    boolean created = db.maybeCreateTable(keyspace, collection);
+    // After creating the table, it can take up to 2 seconds for permissions cache to be updated,
+    // but we can force the permissions refetch by logging in again.
+    if (created) {
+      db = dbFactory.getDocDataStoreForToken(authToken, headers);
+      db.maybeCreateTableIndexes(keyspace, collection);
+    }
+    return db;
+  }
+
+  public List<String> writeManyDocs(
+      String authToken,
+      String keyspace,
+      String collection,
+      InputStream payload,
+      Optional<String> idPath,
+      Db dbFactory,
+      Map<String, String> headers)
+      throws IOException, UnauthorizedException {
+
+    DocumentDB db = dbFactory.getDocDataStoreForToken(authToken, headers);
+    JsonSurfer surfer = JsonSurferGson.INSTANCE;
+
+    db = maybeCreateTableAndIndexes(dbFactory, db, keyspace, collection, headers, authToken);
+    List<String> idsWritten = new ArrayList<>();
+    try (BufferedReader reader = new BufferedReader(new InputStreamReader(payload, "UTF-8"))) {
+      Iterator<String> iter = reader.lines().iterator();
+      ExecutionContext context = ExecutionContext.NOOP_CONTEXT;
+      String docsPath = convertToJsonPtr(idPath.get());
+      int chunkIndex = 0;
+      final int CHUNK_SIZE = 250;
+      while (iter.hasNext()) {
+        chunkIndex++;
+        Map<String, String> docsInChunk = new LinkedHashMap<>();
+        while (docsInChunk.size() < CHUNK_SIZE && iter.hasNext()) {
+          String doc = iter.next();
+          JsonNode json = mapper.readTree(doc);
```

Comment:
Why do we require one doc per line? Is it not possible to parse JSON input in a stream fashion?

Reply:
Technically, the input isn't JSON, since it's actually multiple delimited JSON objects (JSON lines format). I'll look into possible methods to stream JSON lines, but Jackson doesn't have anything for JSON lines afaik

Reply:
Another option would be to make the input valid JSON (i.e. require that it's a JSON array) and then streaming probably becomes possible/easy

Reply:
Yeah, I guess it may depend on how the library works with input streams. Please post your findings :)

Reply:
Requiring input to be a JSON array is not unreasonable from my POV.

Reply:
Perhaps an option to import and/or export in either format?

Reply:
So using a JsonParser (Jackson's impl for streaming JSON) is feasible, but only with valid JSON, so I am going to stick with expecting array data over the wire; then, each object in the array can get parsed one by one using an ObjectMapper, which won't pull in much into memory at once :+1: impl is [here](https://github.com/stargate/stargate/pull/1043/files#diff-caf69767609da0fee30d28b2b410ba5e400ed01e1f42af082f495ef82fec8416R435)